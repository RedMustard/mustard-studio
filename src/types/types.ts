export type KeyboardKeyCount = 88 | 61 | 49 | 37 | 25;
export type KeyboardKeyColor = 'white' | 'black';
export type Octave = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7;
export type OscillatorId = 'osc1' | 'osc2' | 'oscSub';
export type ValueUnit = 'percent' | 'cent' | 'pan';


export type StudioService = {
    settings: {
        master: {
            volume: number,
            pan: number,
        },
        osc1: OscillatorSettings & OscillatorDetune,
        osc2: OscillatorSettings & OscillatorDetune,
        oscSub: OscillatorSettings,
    },
    gainNodes: {
        master: GainNode,
        osc1: GainNode,
        osc2: GainNode,
        oscSub: GainNode,
    },
    panNodes: {
        master: StereoPannerNode,
        osc1: StereoPannerNode,
        osc2: StereoPannerNode,
        oscSub: StereoPannerNode,
    },
};

export type OscillatorSettings = {
    enabled: boolean,
    octave: number,
    pan: number,
    type: OscillatorType,
    volume: number,
};

export type OscillatorDetune = {
    detune: number,
};

export type StudioServiceActionType =
    'SET_MASTER_VOLUME' |
    'SET_MASTER_GAIN' |
    'SET_MASTER_PAN_NODE' |
    'SET_MASTER_PAN_POSITION' |
    'SET_OSC_1_VOLUME' |
    'SET_OSC_2_VOLUME' |
    'SET_OSC_SUB_VOLUME' |
    'SET_OSC_1_GAIN' |
    'SET_OSC_2_GAIN' |
    'SET_OSC_SUB_GAIN' |
    'SET_OSC_1_ENABLED' |
    'SET_OSC_2_ENABLED' |
    'SET_OSC_SUB_ENABLED' |
    'SET_OSC_1_TYPE' |
    'SET_OSC_2_TYPE' |
    'SET_OSC_SUB_TYPE' |
    'SET_OSC_1_DETUNE' |
    'SET_OSC_2_DETUNE' |
    'SET_OSC_1_PAN_NODE' |
    'SET_OSC_2_PAN_NODE' |
    'SET_OSC_SUB_PAN_NODE' |
    'SET_OSC_1_PAN_POSITION' |
    'SET_OSC_2_PAN_POSITION' |
    'SET_OSC_SUB_PAN_POSITION' |
    'SET_OSC_1_OCTAVE' |
    'SET_OSC_2_OCTAVE' |
    'SET_OSC_SUB_OCTAVE';

export type StudioServiceAction = {
    type: StudioServiceActionType;
    payload: any;
};

export type DispatchFunction = (args: StudioServiceAction) => void;

