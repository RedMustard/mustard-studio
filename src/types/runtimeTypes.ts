export enum OscillatorTypeEnum {
    SINE = 'sine',
    SAWTOOTH = 'sawtooth',
    TRIANGLE = 'triangle',
    SQUARE = 'square',
    CUSTOM = 'custom',
}
export const OscillatorTypes = Object.values(OscillatorTypeEnum) as OscillatorType[];

export enum FilterTypeEnum {
    LOWPASS = 'lowpass',
    HIGHPASS = 'highpass',
    BANDPASS = 'bandpass',
    LOWSHELF = 'lowshelf',
    HIGHSHELF = 'highshelf',
    PEAKING = 'peaking',
    NOTCH = 'notch',
    ALLPASS = 'allpass',
}
export const FilterTypes = Object.values(FilterTypeEnum) as BiquadFilterType[];


export enum ValueUnitEnum {
    PERCENT = 'percent',
    CENT = 'cent',
    PAN = 'pan',
    SECONDS = 'seconds',
}

export enum OscillatorIdEnum {
    OSC_1 = 'osc1',
    OSC_2 = 'osc2',
    OSC_SUB = 'oscSub',
}
