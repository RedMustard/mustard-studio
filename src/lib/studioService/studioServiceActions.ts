import { DispatchFunction, OscillatorId } from '../../types/types';
import { logger } from '../utils/logger/logger';

export const setMasterVolume = (value: number, dispatch: DispatchFunction) => {
    dispatch({ type: 'SET_MASTER_VOLUME', payload: value });
    logger.info(`Master volume set with value ${value * 100}%`);
};

export const setMasterGainNode = (gainNode: GainNode, dispatch: DispatchFunction) => {
    dispatch({ type: 'SET_MASTER_GAIN', payload: gainNode });
    logger.info('Master gain node added', gainNode);
};

export const setOscillatorVolume = (value: number, oscillatorId: OscillatorId, dispatch: DispatchFunction) => {
    switch (oscillatorId) {
        case 'osc1':
            dispatch({ type: 'SET_OSC_1_VOLUME', payload: value });
            logger.info(`Oscillator 1 volume set with value ${value * 100}%`);
            break;
        case 'osc2':
            dispatch({ type: 'SET_OSC_2_VOLUME', payload: value });
            logger.info(`Oscillator 2 volume set with value ${value * 100}%`);
            break;
        default:
            logger.info('Unknown oscillatorId provided, no oscillator volume set. Received: ', oscillatorId);
            break;
    }
};

export const setOscillatorGainNode = (gainNode: GainNode, oscillatorId: OscillatorId, dispatch: DispatchFunction) => {
    switch (oscillatorId) {
        case 'osc1':
            dispatch({ type: 'SET_OSC_1_GAIN', payload: gainNode });
            logger.info('Oscillator 1 gain node added', gainNode);
            break;
        case 'osc2':
            dispatch({ type: 'SET_OSC_2_GAIN', payload: gainNode });
            logger.info('Oscillator 1 gain node added', gainNode);
            break;
        default:
            logger.info('Unknown oscillatorId provided, no oscillator gain node set. Received: ', oscillatorId);
            break;
    }
};

export const setOscillatorEnabled = (isEnabled: boolean, oscillatorId: OscillatorId, dispatch: DispatchFunction) => {
    switch (oscillatorId) {
        case 'osc1':
            dispatch({ type: 'SET_OSC_1_ENABLED', payload: isEnabled });
            logger.info('Oscillator node 1 isEnabled: ', isEnabled);
            break;
        case 'osc2':
            dispatch({ type: 'SET_OSC_2_ENABLED', payload: isEnabled });
            logger.info('Oscillator node 1 isEnabled: ', isEnabled);
            break;
        default:
            logger.info('Unknown oscillatorId provided, no oscillator set. Received: ', oscillatorId);
            break;
    }
};

export const setOscillatorType = (oscillatorType: OscillatorType, oscillatorId: OscillatorId, dispatch: DispatchFunction) => {
    switch (oscillatorId) {
        case 'osc1':
            dispatch({ type: 'SET_OSC_1_TYPE', payload: oscillatorType });
            logger.info(`Oscillator node 1 type set to ${oscillatorType}`);
            break;
        case 'osc2':
            dispatch({ type: 'SET_OSC_2_TYPE', payload: oscillatorType });
            logger.info(`Oscillator node 2 type set to ${oscillatorType}`);
            break;
        default:
            break;
    }
};

export const setOscillatorDetune = (value: number, oscillatorId: OscillatorId, dispatch: DispatchFunction) => {
    switch (oscillatorId) {
        case 'osc1':
            dispatch({ type: 'SET_OSC_1_DETUNE', payload: value });
            logger.info(`Oscillator node 1 detune set to ${value}`);
            break;
        case 'osc2':
            dispatch({ type: 'SET_OSC_2_DETUNE', payload: value });
            logger.info(`Oscillator node 2 detune set to ${value}`);
            break;
        default:
            break;
    }
};
