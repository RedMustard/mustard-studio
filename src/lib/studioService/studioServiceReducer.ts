import { GenericAction, StudioService } from '../../types/types';

export const studioServiceReducer = (state: StudioService, action: GenericAction) => {
    switch (action.type) {
        case 'SET_MASTER_VOLUME':
            return {
                ...state,
                volume: {
                    ...state.volume,
                    master: action.payload,
                },
            };

        case 'SET_OSCILLATOR_1_VOLUME':
            return {
                ...state,
                volume: {
                    ...state.volume,
                    oscillator1: action.payload,
                },
            };

        case 'SET_OSCILLATOR_2_VOLUME':
            return {
                ...state,
                volume: {
                    ...state.volume,
                    oscillator2: action.payload,
                },
            };

        case 'SET_MASTER_GAIN':
            return {
                ...state,
                gainNodes: {
                    ...state.gainNodes,
                    master: action.payload,
                },
            };

        case 'SET_OSCILLATOR_1_GAIN':
            return {
                ...state,
                gainNodes: {
                    ...state.gainNodes,
                    oscillator1: action.payload,
                },
            };

        case 'SET_OSCILLATOR_2_GAIN':
            return {
                ...state,
                gainNodes: {
                    ...state.gainNodes,
                    oscillator2: action.payload,
                },
            };

        default:
            return state;
    }
};
