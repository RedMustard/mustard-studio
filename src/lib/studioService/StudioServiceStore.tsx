import { createContext, h } from 'preact';
import { useReducer } from 'preact/hooks';
import { OSC_1_INITIAL_SETTINGS, OSC_2_INITIAL_SETTINGS, OSC_SUB_INITIAL_SETTINGS } from '../../constants';
import { DispatchFunction, StudioService } from '../../types/types';
import { deepClone } from '../utils/objects/objects';
import { studioServiceReducer } from './studioServiceReducer';


const initialState: StudioService = Object.freeze({
    master: {
        gainNode: undefined,
        panNode: undefined,
        settings: {
            volume: 0.1,
            pan: 0,
        },
    },
    oscillators: {
        osc1: {
            gainNode: undefined,
            panNode: undefined,
            settings: {
                ...OSC_1_INITIAL_SETTINGS,
            },
        },
        osc2: {
            gainNode: undefined,
            panNode: undefined,
            settings: {
                ...OSC_2_INITIAL_SETTINGS,
            },
        },
        oscSub: {
            gainNode: undefined,
            panNode: undefined,
            settings: {
                ...OSC_SUB_INITIAL_SETTINGS,
            },
        },
    },
});

export const getInitialState = () => deepClone(initialState);

export const StudioServiceContext = createContext<[StudioService, DispatchFunction]>([getInitialState(), () => {}]);

export const StudioServiceStore = ({ children }: any) => {
    const [state, dispatch]: [StudioService, DispatchFunction] = useReducer(studioServiceReducer, getInitialState());

    return (
        <StudioServiceContext.Provider value={[state, dispatch]}>
            {children}
        </StudioServiceContext.Provider>
    );
};
