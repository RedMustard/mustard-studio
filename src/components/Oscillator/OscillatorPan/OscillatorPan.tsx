import { h } from 'preact';
import { useContext, useEffect } from 'preact/hooks';
import { MAX_PAN, MIN_PAN, PAN_STEPS } from '../../../constants';
import { resetOscillatorPanPosition, setOscillatorPanNode, setOscillatorPanPosition } from '../../../lib/studioService/studioServiceActions';
import { StudioServiceContext } from '../../../lib/studioService/StudioServiceStore';
import { ValueUnitEnum } from '../../../types/runtimeTypes';
import { OscillatorId } from '../../../types/types';
import { Fader } from '../../Fader/Fader';

interface OscillatorPanProps {
    audioContext: AudioContext;
    oscillatorId: OscillatorId;
}

export const OscillatorPan = ({ audioContext, oscillatorId }: OscillatorPanProps) => {
    const [studioService, dispatch] = useContext(StudioServiceContext);
    const { oscillators, master } = studioService;
    const oscillatorPanPosition = oscillators[oscillatorId].settings.pan;
    const oscillatorGainNode = oscillators[oscillatorId].gainNode;
    const masterGainNode = master.gainNode;
    let oscillatorPanNode = oscillators[oscillatorId].panNode;

    const handleOnMouseDown = (e: MouseEvent) => {
        if (e.buttons === 1 && (e.ctrlKey || e.metaKey)) {
            resetOscillatorPanPosition(oscillatorId, dispatch);
        }
    };

    useEffect(() => {
        if (!oscillatorPanNode) {
            oscillatorPanNode = audioContext.createStereoPanner();
            setOscillatorPanNode(oscillatorPanNode, oscillatorId, dispatch);
        } else {
            oscillatorPanNode.pan.value = oscillatorPanPosition;
            oscillatorPanNode.connect(masterGainNode);
            oscillatorGainNode.connect(oscillatorPanNode);
        }
    }, [oscillatorPanNode, oscillatorPanPosition, oscillatorGainNode, masterGainNode]);

    return (
        <div className="oscillator-pan">
            Pan: &nbsp;
            <Fader
                classSuffix="oscillator-pan"
                value={oscillatorPanPosition}
                onInput={(value) => setOscillatorPanPosition(value, oscillatorId, dispatch)}
                onMouseDown={(e: MouseEvent) => handleOnMouseDown(e)}
                maxValue={MAX_PAN}
                minValue={MIN_PAN}
                stepResolution={PAN_STEPS}
                valueUnit={ValueUnitEnum.PAN}
            />
        </div>
    );
};
