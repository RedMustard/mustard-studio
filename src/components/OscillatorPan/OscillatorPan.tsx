import { h } from 'preact';
import { useContext, useEffect } from 'preact/hooks';
import { MAX_PAN, MIN_PAN, PAN_STEPS } from '../../constants';
import { setOscillatorPanNode, setOscillatorPanPosition } from '../../lib/studioService/studioServiceActions';
import { StudioServiceContext } from '../../lib/studioService/StudioServiceStore';
import { ValueUnitEnum } from '../../types/runtimeTypes';
import { OscillatorId } from '../../types/types';
import { Fader } from '../Fader/Fader';

interface OscillatorPanProps {
    audioContext: AudioContext;
    oscillatorId: OscillatorId;
}

export const OscillatorPan = ({ audioContext, oscillatorId }: OscillatorPanProps) => {
    const [studioService, dispatch] = useContext(StudioServiceContext);
    const { settings, gainNodes, panNodes } = studioService;
    const oscillatorPanPosition = settings[oscillatorId].pan;
    const oscillatorGainNode = gainNodes[oscillatorId];
    const masterGainNode = gainNodes.master;
    let oscillatorPanNode = panNodes[oscillatorId];

    useEffect(() => {
        if (!oscillatorPanNode) {
            oscillatorPanNode = audioContext.createStereoPanner();
            setOscillatorPanNode(oscillatorPanNode, oscillatorId, dispatch);
        } else {
            oscillatorPanNode.pan.value = oscillatorPanPosition;
            oscillatorPanNode.connect(masterGainNode);
            oscillatorGainNode.connect(oscillatorPanNode);
        }
    }, [oscillatorPanNode, oscillatorPanPosition, oscillatorGainNode, masterGainNode]);

    return (
        <div className="oscillator-pan">
            Pan: &nbsp;
            <Fader
                classSuffix="oscillator-pan"
                value={oscillatorPanPosition}
                onInput={(value) => setOscillatorPanPosition(value, oscillatorId, dispatch)}
                maxValue={MAX_PAN}
                minValue={MIN_PAN}
                stepResolution={PAN_STEPS}
                valueUnit={ValueUnitEnum.PAN}
            />
        </div>
    );
};
