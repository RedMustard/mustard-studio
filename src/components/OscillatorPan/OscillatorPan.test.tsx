import { h } from 'preact';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { OscillatorPan } from './OscillatorPan';
import { setOscillatorPanPosition, setOscillatorPanNode, resetOscillatorPanPosition } from '../../lib/studioService/studioServiceActions';
import { getInitialState, StudioServiceContext } from '../../lib/studioService/StudioServiceStore';
import { OscillatorId } from '../../types/types';


jest.mock('../../lib/studioService/studioServiceActions');

// eslint-disable-next-line import/newline-after-import
const wamock = require('web-audio-mock-api');
const audioContext = new wamock.AudioContext();
const baseProps = {
    audioContext,
    oscillatorId: 'osc1' as OscillatorId,
};
const mockPanNode = {
    connect: jest.fn().mockReturnThis(),
    pan: {
        value: 1.0,
    },
};
const mockGainNode = {
    connect: jest.fn().mockReturnThis(),
    gain: {
        value: 1.0,
    },
};
const initialState = {
    ...getInitialState(),
    oscillators: {
        osc1: {
            panNode: mockPanNode,
            gainNode: mockGainNode,
            settings: {
                pan: 0.5,
            },
        },
    },
};

describe('<OscillatorPan />', () => {
    beforeEach(() => {
        jest.spyOn(audioContext, 'createStereoPanner').mockImplementation(() => mockPanNode);
    });

    it('renders with basic props', () => {
        const wrapper = shallow(
            <OscillatorPan
                {...baseProps}
            />,
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('calls setOscillatorPanNode if oscillatorPanNode does not exist', () => {
        shallow(
            <OscillatorPan
                {...baseProps}
            />,
        );
        expect(audioContext.createStereoPanner).toBeCalled();
        expect(setOscillatorPanNode).toBeCalled();
    });

    it('calls setOscillatorPanPosition when pan changed', () => {
        const wrapper = mount(
            <OscillatorPan
                {...baseProps}
            />,
        );
        wrapper.find('input').simulate('input');
        expect(setOscillatorPanPosition).toBeCalled();
    });

    it('sets mockPanNode.pan.value, connects mockPanNode to master gain node if OscillatorPanNode exists', () => {
        mount(
            <StudioServiceContext.Provider value={[initialState, jest.fn()]}>
                <OscillatorPan
                    {...baseProps}
                />
            </StudioServiceContext.Provider>,
        );
        expect(mockGainNode.connect).toBeCalled();
        expect(mockPanNode.connect).toBeCalled();
        expect(mockPanNode.pan.value).toBe(initialState.oscillators.osc1.settings.pan);
    });

    it('calls resetOscillatorPanPosition when ctrl clicked', () => {
        const wrapper = mount(
            <OscillatorPan
                {...baseProps}
            />,
        );
        wrapper.find('input').simulate('mousedown', { buttons: 1, ctrlKey: true });
        expect(resetOscillatorPanPosition).toBeCalled();
    });

    it('calls resetOscillatorPanPosition when cmd clicked', () => {
        const wrapper = mount(
            <OscillatorPan
                {...baseProps}
            />,
        );
        wrapper.find('input').simulate('mousedown', { buttons: 1, metaKey: true });
        expect(resetOscillatorPanPosition).toBeCalled();
    });

    it('does not call resetOscillatorPanPosition when not ctrl or cmd clicked', () => {
        const wrapper = mount(
            <OscillatorPan
                {...baseProps}
            />,
        );
        wrapper.find('input').simulate('mousedown', { buttons: 1 });
        expect(resetOscillatorPanPosition).toBeCalled();
    });
});
