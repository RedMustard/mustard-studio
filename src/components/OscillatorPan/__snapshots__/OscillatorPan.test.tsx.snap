// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OscillatorPan /> renders with basic props 1`] = `
<div
  className="oscillator-pan"
>
  Pan: Â 
  <Fader
    classSuffix="oscillator-pan"
    maxValue={1}
    minValue={-1}
    onInput={[Function]}
    onMouseDown={[Function]}
    stepResolution={0.01}
    value={0}
    valueUnit="pan"
  />
</div>
`;
